### PR TITLE
HPCC-13275 Split the child dependency spotting from the resourcer

### PR DESCRIPTION
**Pull request by ghalliday**
This change is part of the preparation for improving the execution of conditional
expressions.  By splitting the child dataset spotting from the resourcing it
allows other transformations to also be applied within graph "resourcing".

The change looks substantial, but some of that is moving code and refactoring it.
The main changes are:
- Split the child dataset spotting into a separate stage
- Move some logic into the the options class which is shared between stages
- Move the dependency tracking code into a separate class, and reduce the lifetime
  of some of the variables.
- Allow graph results to be reused by the resourcer
- Introduce the concept of a _graphLocal_ action which is only executed when required
  by another activity.

Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman @jamienoss please review.

This change also has the side-effect of fixing https://track.hpccsystems.com/browse/HPCC-10105.

Note - hqlresource.cpp contains too many changes so it isn't include in the github diff.  You'll need to directly look at the change.  You may benefit from using a comparison program to match the code that has moved classes to reduce the size of the code to check.

It is going to be hard to review this.  I have added some regression tests to ensure that conditional expressions are still only executed when required.  The changes in the public regression suites are minimal - but please compare them yourself.  

There are more changes in the private suite - almost all improvements - most come from fewer graph results being used.  I have looked at a random selection and they all seem good.

File: ecl/hqlcpp/hqlresource.ipp [hunk 1]
```diff
@@ -393,34 +408,20 @@ public:
...
     HqlExprAttr pathToSplitter;
     HqlExprArray aggregates;
     HqlExprArray conditions;
-    ChildDependentArray childDependents;
-    HqlExprAttr spilledDataset;
     HqlExprAttr splitterOutput;
     HqlExprArray projected;
     HqlExprAttr projectedExpr;
     CIArrayOf<CSplitterLink> balancedLinks;
+    GraphLinkArray dependsOn;           // NB: These do no link....
```
Comment (jamienoss) [inline ecl/hqlcpp/hqlresource.ipp]: Typo - ...do no**T** link...

**Comment by jamienoss** [comment]
hqlresource.cpp:6604 "(?? is this needed)" is this a personal dev comment or a relevant one to remain?

**Comment by jamienoss** [comment]
Additionally 6620-21 lack of and position of '.'.

**Comment by jamienoss** [comment]
@ghalliday I have now completed one full pass of both the code and the reg suit results. The former being easier than the latter to review. There is a lot of refactoring that is not so difficult to follow, and most of the changes I could also follow. In all, this seems sound. I would, however, like to make another pass just to look into a few things that I was unsure of. I should have this complete by the end of tomorrow.

File: ecl/hqlcpp/hqlresource.ipp (hunk 1 above)
Comment (ghalliday) [inline ecl/hqlcpp/hqlresource.ipp]: ok, will also fix 207 - where it was copied from

**Comment by ghalliday** [comment]
6604 - "is this needed".  Yes it is still relevant to retain - comment updated.

**Comment by ghalliday** [comment]
@jamienoss changes so far in a second commit

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success

**Comment by jamienoss** [comment]
@ghalliday, @richardkchapman Walked over it again, and finally, this looks good to me to be merged.

**Pull request merged**